README.md
# Old/archived codebase

**This codebase is old/deprecated and probably would need updating to run today**. It is here for archival/nostalgia purposes.

# oerperl and oerperl+MySQL

This project adds Perl-scripting support to the oer and oer+MySQL IRC bots.

Homepage: https://vanhat.rikunmediatuotanto.fi/oerperl/

oer and oer+MySQL are copyright 2000-2005 Samuli "EQU" Kaski

https://web.archive.org/web/20160925102343/http://oer.equnet.org/?disclaimer=shown
http://www.kaski-net.net/

# License

Copyright (c) 2004, 2005 Riku Nurminen

https://www.nurminen.dev

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.


CHANGELOG
1.0-42
------

- Rewrote the IRC client protocol parser due the PART issue
  discovered in -41 (this change could break just about
  anything!)
- Improved the channel mode, channel key and channel limit
  parsing in oer+MySQL (there is still room for improvement)
- Improved the socket family detection so that oer+MySQL
  should no longer attempt to connect to a IPv4 address from
  a IPv6 address and visa versa
  (Thanks must go to Renzo)
- Scheduled modes weren't adjusted when there was a nick
  change for a nick for which a scheduled mode existed, fixed
- Fixed yet another IPv4 <-> IPv6 interoperatibility issue
  (Thanks must go to Emperor)
- There was a missplaced freeaddrinfo() in network.c
  (Thanks must go to Lasse Lehtinen)
- Added Operating System information to CTCP VERSION if
  supported (shown only to bot users)
- Configure now checks for the availability of getaddrinfo()
- dbconnect didn't actually reconnect to the database since
  mysql_ping() doesn't automatically attempt a reconnection
  (Thanks must go to Sekto)
- dbconnect didn't check all MySQL connections, fixed
- oer+MySQL will now attempt to re-establish a broken MySQL
  connection in 10 minute intervals, not 30
- When the database connection has been disconnected oer+MySQL
  will no longer attempt to clean expired autheds until the
  database connection is available again
- I fixed a long lasting error in where oer+MySQL would
  sometimes get stuck in the server registration phase
- Added notice flood protection (how on earth have I
  overlooked this one)
- oer+MySQL would exit if it couldn't resolve the server
  hostname, fixed
- Changed gethostbyname() references to getaddrinfo()
- ECONNRESET in connect() is no longer seen as a fatal error
- Implemented reset autheds, requires global admin authorization
- Implemented reset masks, requires global admin authorization
  (Thanks must go to Refic)
- cleanautheds() would flood shared bots if more than one
  authed was cleaned, fixed
- Updated copyright strings to include year 2004

1.0-41
------

- The PART message has changed in ircu (the Undernet ircd)
  and it's derivatives (which include for example the ircd
  used in QuakeNet) -- Until I sort out whether this is
  indeed a bug or a known change in the newer ircu's here
  is a fix for it
  (Thanks must go to Emperor)
- Added uptime in seconds to CTCP VERSION
- Logoff didn't support multiple arguments, fixed
- Updated copyright strings to include 2003
- Added IPv6 support
  (Initial patch provided by Rakkis / MadQon, Thanks!)
- edit and del chankey returned nothing to the user, fixed
- dop, devoice, op and voice didn't check that the target
  nick was on channel before issuing the MODE, fixed

1.0-40
------

- A major oversight happened in -38 when I rewrote the
  authentication checks in parse.c, fixed
  (Thanks must go to Emperor)

1.0-39
------

- clearbans no longer removes permbans
  (Thanks must go to many out there, refic for reminding me)
- There was a typo in requestop, fixed (4 -> 5)
  (Thanks must go to Jase)
- Increased the maximum length of a single topic from 160
  to 250 characters, maximum number of topics is still 20
  (Thanks must go to Jase)
- There was a bug in the topic parser, fixed
  (Thanks must go to Jase)
- edit server was broken, fixed
  (Thanks must go to Renzo)
- jump didn't work as expected, fixed
- Fixed several commands which would eventuelly at some point
  corrupt their linked list structures:
	+ delnickbks()
	+ delwordbks()
	+ deladmin()
	+ deladminmask()
	+ deluser()
	+ delusermask()
	+ deltrusted()
	+ delservice()
	+ delchannel()
	+ deladverts()
	+ delserver()
	+ delchannelban()
	+ timed_del()
	+ mmode_del()
	+ userleft()
	+ cleanautheds()
	+ freepubmsguser()
	+ freenickchanges()
- initmmodes() left the list non null, fixed
- initmmodesfornick() left the list non null even if the
  list became empty cause of the operation, fixed
- Did some code clean up
- The change in -38 which makes oer+MySQL assume that a user
  which has @ also haves +v was broken, fixed
  (Thanks must go to Emperor)
- There were some issues with the detection of a broken
  database connection, fixed
- The manual reconnection to a database wasn't functioning
  properly, fixed
- You can now specify the full path of the config with -f
  (increased the length of the variable)
- validateflags() sometimes returned bogus values, fixed
- oer_doquery() no longer logs the MySQL execution times,
  the change was necessary due to a MySQL limitation (if
  output logging was switched on it would result to a
  "Commands out of sync" error message, the other way to
  fix this would be to have a second database connection
  just for output logging, IMO it's not worth it)
- Added global flag "S" to allow a shared oer+MySQL setup
  (in other words you will need to set global flag "S" to
  all your shared oer+MySQL beginning with this version if
  you require a shared setup, default is off)
- Topic delete now allows you to delete several topics
  at a time, for example "!td 2 5 7"
  (Thanks must go to refic)
- Removed the annoying "channel key is now.." and "deleted
  channel key from.." messages
  (Thanks must go to Renzo)
- Removed the custom isnumber() function (oer+MySQL wouldn't
  compile on FreeBSD)
  (Thanks must go to Trollo)
- Added SYNCGUIDE which you should read if you plan to
  implement a shared oer+MySQL setup
- Now admins and users can be fetched from a different
  database/MySQL installation (see SYNCGUIDE)
- tools/changeident.pl and tools/delident.pl were missing
  oer_ext table, fixed
- Added tools/delsyncdata.pl
- A friend (user flag "f") is now also allowed to use prohibited
  nicks, was the only thing a friend wasn't allowed to do
  (Thanks must go to Renzo)

1.0-38
------

- Added a timer to adverts just like with on join quotes,
  adverts on a busy channel were sent to each person that
  joined the channel, everytime (would cause "Target changing
  too fast" errors on heavy loaded bots and lead to adverts

reg.c
#include <stdio.h>
#include <ctype.h>

/*
 * Written By Douglas A. Lewis <dalewis@cs.Buffalo.EDU>
 *
 * This file is in the public domain.
 */

/* modified by EQU to use glibc tolower() */

#define RETURN_FALSE -1
#define RETURN_TRUE count

char lower_tab[256] = 
{
  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,
 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63,
 64, 97, 98, 99,100,101,102,103,104,105,106,107,108,109,110,111,
112,113,114,115,116,117,118,119,120,121,122, 91, 92, 93, 94, 95,
 96, 97, 98, 99,100,101,102,103,104,105,106,107,108,109,110,111,
112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,
128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,
144,145,145,147,148,149,150,151,152,153,154,155,156,157,158,159,
160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,
176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,
192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,
208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,
224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,
240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255
};

#if 0
#undef tolower		/* don't want previous version. */
#define tolower(x) lower_tab[x]
#endif

/*
 * So I don't make the same mistake twice: We don't need to check for '\\'
 * in the input (string to check) because it doesn't have a special meaning
 * in that context.
 *
 * We don't need to check for '\\' when we come accross a * or % .. The rest
 * of the routine will do that for us.
 */

static int _wild_match(char *mask, char *string)
{
	char	*m = mask,
		*n = string,
		*ma = NULL,
		*na = NULL,
		*mp = NULL,
		*np = NULL;
	int	just = 0,
		pcount = 0,
		acount = 0,
		count = 0;

	for (;;)
	{
		if (*m == '*')
		{
			ma = ++m;
			na = n;
			just = 1;
			mp = NULL;
			acount = count;
		}
		else if (*m == '%')
		{
			mp = ++m;
			np = n;
			pcount = count;
		}
		else if (*m == '?')
		{
			m++;
			if (!*n++)
				return RETURN_FALSE;
		}
		else
		{
			if (*m == '\\')
			{
				m++;
				/* Quoting "nothing" is a bad thing */
				if (!*m)
					return RETURN_FALSE;
			}
			if (!*m)
			{
				/*
				 * If we are out of both strings or we just
				 * saw a wildcard, then we can say we have a
				 * match
				 */
				if (!*n)
					return RETURN_TRUE;
				if (just)
					return RETURN_TRUE;
				just = 0;
				goto not_matched;
			}
			/*
			 * We could check for *n == NULL at this point, but
			 * since it's more common to have a character there,
			 * check to see if they match first (m and n) and
			 * then if they don't match, THEN we can check for
			 * the NULL of n
			 */
			just = 0;
			if (tolower((int)*m) == tolower((int)*n))
			{
				m++;
				if (*n == ' ')
					mp = NULL;
				count++;
				n++;
			}
			else
			{

	not_matched:

				/*
				 * If there are no more characters in the
				 * string, but we still need to find another
				 * character (*m != NULL), then it will be
				 * impossible to match it
				 */
				if (!*n)
					return RETURN_FALSE;
				if (mp)
				{
					m = mp;
					if (*np == ' ')
					{
						mp = NULL;
						goto check_percent;
					}
					n = ++np;
					count = pcount;
				}
				else
	check_percent:

				if (ma)
				{
					m = ma;
					n = ++na;
					count = acount;
				}
				else
					return RETURN_FALSE;
			}
		}
	}
}

int wild_match(char *pattern, char *str)
{
        /* assuming a -1 return of false */
	return _wild_match((char *) pattern, (char *) str) + 1;
}
